pkcs11/src/new/types/object.rs

```diff
@@ -171,27 +235,133 @@ impl Attribute<'_> {
             | Attribute::Unwrap(b)
             | Attribute::Verify(b)
             | Attribute::VerifyRecover(b)
-            | Attribute::Wrap(b) => *b as *mut _ as CK_VOID_PTR,
-            Attribute::Base(byte) => *byte as *mut _ as CK_VOID_PTR,
-            Attribute::Class(object_class) => *object_class as CK_OBJECT_CLASS_PTR as CK_VOID_PTR,
-            Attribute::KeyType(key_type) => *key_type as *mut _ as CK_VOID_PTR,
-            Attribute::Label(label) => label.as_ptr() as CK_VOID_PTR,
-            Attribute::ModulusBits(bits) => *bits as *mut _ as CK_VOID_PTR,
-            Attribute::Prime(bytes) => bytes.as_ptr() as CK_VOID_PTR,
-            Attribute::PublicExponent(bytes) => bytes.as_ptr() as CK_VOID_PTR,
-            Attribute::Modulus(bytes) => bytes.as_ptr() as CK_VOID_PTR,
-            Attribute::Value(value) => value.as_ptr() as CK_VOID_PTR,
-            Attribute::ValueLen(len) => *len as *mut _ as CK_VOID_PTR,
+            | Attribute::Wrap(b) => b.as_ref().get_ref() as *const _ as *mut c_void,
+            // CK_ULONG
+            Attribute::ModulusBits(val) | Attribute::ValueLen(val) => {
+                val.as_ref().get_ref() as *const _ as *mut c_void
+            }
+            // Vec<u8>
+            Attribute::Base(bytes)
+            | Attribute::Label(bytes)
+            | Attribute::Prime(bytes)
+            | Attribute::PublicExponent(bytes)
+            | Attribute::Modulus(bytes)
+            | Attribute::Value(bytes)
+            | Attribute::Id(bytes) => bytes.as_ptr() as *mut c_void,
+            // Unique types
+            Attribute::Class(object_class) => {
+                object_class.as_ref().get_ref() as *const _ as *mut c_void
+            }
+            Attribute::KeyType(key_type) => key_type.as_ref().get_ref() as *const _ as *mut c_void,
+            Attribute::AllowedMechanisms(mechanisms) => {
+                mechanisms.as_ref().get_ref() as *const _ as *mut c_void
+            }
         }
     }
 }
 
-impl From<&mut Attribute<'_>> for CK_ATTRIBUTE {
-    fn from(attribute: &mut Attribute) -> Self {
+impl From<&Attribute> for CK_ATTRIBUTE {
+    fn from(attribute: &Attribute) -> Self {
         Self {
             type_: attribute.attribute_type().into(),
             pValue: attribute.ptr(),
-            ulValueLen: attribute.len(),
+            // Truncation from usize to CK_ULONG not checked
+            // Should be fine in most cases.
```

Comment:
If I don't do this cast, I have to use `TryFrom` instead which is a bit annoying...
Maybe here is a good place to panic?

Reply:
Yes, I would rather we panic than failing silently. Unless you put a huge amount of data in an attribute it won't fail, right? Given a good panic message, at least the source can be found/understood.